src/query/sql/src/planner/optimizer/heuristic/heuristic.rs

```diff
@@ -113,13 +121,27 @@ impl HeuristicOptimizer {
         Ok(result)
     }
 
+    fn calc_operator_rule_set(&self, operator: &RelOperator) -> roaring::RoaringBitmap {
+        unsafe { operator.transformation_candidate_rules() & (&RULE_FACTORY.transformation_rules) }
+    }
+
+    fn get_rule(&self, rule_id: u32) -> Result<RulePtr> {
+        unsafe {
+            RULE_FACTORY.create_rule(
+                DEFAULT_REWRITE_RULES[rule_id as usize],
+                Some(self.metadata.clone()),
+            )
+        }
+    }
+
     /// Try to apply the rules to the expression.
     /// Return the final result that no rule can be applied.
     fn apply_transform_rules(&self, s_expr: &SExpr) -> Result<SExpr> {
         let mut s_expr = s_expr.clone();
+        let rule_set = self.calc_operator_rule_set(&s_expr.plan);
```

Comment:
We should distinguish between `RuleSet` and `RuleList`. `RuleSet` is order-insensitive, and `RuleList` is order-sensitive.

In this case, a `RuleList` is required.

Reply:
> We should distinguish between `RuleSet` and `RuleList`. `RuleSet` is order-insensitive, and `RuleList` is order-sensitive.
> 
> In this case, a `RuleList` is required.

Please refer to the introduction of my implementation ideas above. If you use RuleList, if you can quickly calculate the rules that can be applied by each operator, it will cause too many loops and reduce performance.

Reply:
@leiysky 

Any other question？

Reply:
> And Please refer to the benchmark results

Could you please provide more details about the benchmark? The benefits appear to be relatively low.

hits: from 1.01 to 0.99

![image](https://user-images.githubusercontent.com/5351546/226771851-577467c1-4170-4179-8662-9434c04a25b5.png)

tpch: from 1.02 to 0.98

![image](https://user-images.githubusercontent.com/5351546/226771914-0cb909af-a1a1-4f62-9809-52405412a547.png)

Although performance is important, I prefer to prioritize code readability and maintainability. It's not worth sacrificing those aspects for the sake of minor performance improvements.

Reply:
> 

I disagree with this point of view. With the increase of rules, each rule will increase the number of cycles optimized by unrelated operators, and it is exponential. When I think of it, the performance comparison will be more obvious.

Reply:
> With the increase of rules, each rule will increase the number of cycles optimized by unrelated operators, and it is exponential.

You are correct, but for now, it is not worth it. We can optimize when we identify the bottleneck.